### PR TITLE
[PDS-253270] Treat matching tokens and number of hosts across token ranges as sufficient for consistency

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
At k8s deployments we find that the IPs of nodes change over time. This leads to a race condition where if this change happens while the cluster is live reloading, we incorrectly declare an inconsistent ring when the situation is completely benign.
==COMMIT_MSG==
Accept matching token ranges and number of hosts associated with each range as sufficient for having a consistent view of the token ring. Previously, there was a race condition at k8s deployments that meant that we would declare an inconsistent ring when this was not the case.
==COMMIT_MSG==

**Implementation Description (bullets)**:
- Extract only the start token, end token and number of endpoints from each token range, and use these for comparison.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Added new tests

**Concerns (what feedback would you like?)**:
Is this too generous?

**Where should we start reviewing?**:
TokenRangeResolution.java

**Priority (whenever / two weeks / yesterday)**: Next week Tuesday please.

@tpetracca / @Sam-Kramer for SA

### CODE REVIEW

**Review by schlosna — APPROVED** [review]

**Review by sudiksha27 — COMMENTED** [review]
Thanks for the fix! Did a cursory pass, have a question.

**Review by Jolyon-S — COMMENTED** [review]
I'm pretty confident that this PR does what it claims to do. The question I have is that are we happy with what this does? For example, we'll approve this:

[start: X, end: Y, endpoint: A | start: U, end: V, endpoint: B]
[start: X, end: Y, endpoint B | start U, end: V, endpoint: A]

Do we expect the hosts to be able to shuffle entirely, or just become new (i.e. in the above, only accept if the endpoints differ and we don't have any straight up reassigned hosts)?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/TokenRangeResolution.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            log.trace("<= 1 distinct views of token ranges were provided, so these must have consistent endpoints.");
+            return true;
+        }
+
+        Set<IdentityAgnosticRanges> distinctIdentityAgnosticRanges = tokenRangeViews.stream()
+                .map(ranges -> ranges.stream()
+                        .map(IdentityAgnosticRange::fromTokenRange)
+                        .collect(Collectors.toList()))
+                .map(ImmutableIdentityAgnosticRanges::of)
+                .collect(Collectors.toSet());
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/TokenRangeResolution.java]: Makes sense, I don't think we care particularly about the orderings (and this is fixed in @schlosna's PR).

**Comment by jeremyk-91** [comment]
> [start: X, end: Y, endpoint: A | start: U, end: V, endpoint: B]
> [start: X, end: Y, endpoint B | start U, end: V, endpoint: A]

I think we accept shuffling, as the bad case here is that some nodes think A has a view of the token ring that exactly matches what other nodes think B has, which is really tough to do - even if such an outdated view did exist, it would require B to perfectly match A's tokens which in a world with vnodes is exceedingly unlikely.

**Review by sudiksha27 — APPROVED** [review]

**Pull request merged**